include/ginkgo/core/base/utils_helper.hpp

```diff
@@ -328,6 +372,43 @@ inline const typename std::decay<T>::type* as(const U* obj)
 }
 
 
+/**
+ * Performs polymorphic type conversion on a pointer_param.
+ *
+ * @tparam T  requested result type
+ * @tparam U  static type of the passed object
+ *
+ * @param obj  the object which should be converted
+ *
+ * @return If successful, returns a pointer to the subtype, otherwise throws
+ *         NotSupported.
+ */
+template <typename T, typename U>
+inline typename std::decay<T>::type* as(pointer_param<U> obj)
```

Comment:
we only accept the unique_ptr<U>&& in the above line for the clear ownership.
Won't it destory the unique_ptr case?
pass unique_ptr and implicitly convert to pointer_param and get the raw pointer
It also gives the different type not like the other case return the same type (shared->shared, raw->raw..)

Reply:
This will never happen, because unique_ptr isn't an exact match for pointer_param, thus template argument deduction will fail.

Reply:
you are right. pointer_param always requires the <U> which is not from the constructor for unique_ptr.